dotnet/src/InternalUtilities/src/Diagnostics/ModelDiagnostics.cs

```diff
@@ -0,0 +1,246 @@
+    /// The `formatCompletions` delegate won't be invoked if events are disabled.
+    /// </summary>
+    private static void SetCompletionResponse<T>(
+        Activity activity,
+        T completions,
+        int? promptTokens,
+        int? completionTokens,
+        Func<T, string> formatCompletions) where T : IEnumerable<KernelContent>
+    {
+        if (!IsModelDiagnosticsEnabled())
+        {
+            return;
+        }
+
+        activity.AddTags(
+            [
+                new(ModelDiagnosticsTags.FinishReason, GetFinishReasons(completions)),
+                new(ModelDiagnosticsTags.PromptToken, promptTokens ?? -1),
+                new(ModelDiagnosticsTags.CompletionToken, completionTokens ?? -1),
+                new(ModelDiagnosticsTags.ResponseId, GetResponseId(completions.FirstOrDefault())),
+            ]);
+
+        if (s_enableSensitiveEvents)
+        {
+            activity.AttachSensitiveDataAsEvent(
+                ModelDiagnosticsTags.CompletionEvent,
+                [
+                    new(ModelDiagnosticsTags.CompletionEventCompletion, formatCompletions(completions)),
+                ]);
+        }
+    }
+
+    private static string GetFinishReasons(IEnumerable<KernelContent> completions)
+    {
+        var finishReasons = completions.Select(c =>
+        {
+            if (c.Metadata?.TryGetValue("FinishReason", out var finishReason) == true)
+            {
+                return finishReason;
+            }
+
+            return null;
+        });
+
+        return $"[{string.Join(", ", finishReasons)}]";
+    }
+
+    private static string GetResponseId(KernelContent? completion)
+    {
+        if (completion?.Metadata?.TryGetValue("Id", out var id) == true)
+        {
+            return id as string ?? "N/A";
+        }
+
+        return "N/A";
+    }
+
+    /// <summary>
+    /// Tags used in model diagnostics
+    /// </summary>
+    private static class ModelDiagnosticsTags
+    {
+        // Activity tags
+        public const string System = "gen_ai.system";
+        public const string Operation = "gen_ai.operation.name";
+        public const string Model = "gen_ai.request.model";
+        public const string MaxToken = "gen_ai.request.max_token";
+        public const string Temperature = "gen_ai.request.temperature";
+        public const string TopP = "gen_ai.request.top_p";
+        public const string ResponseId = "gen_ai.response.id";
+        public const string ResponseModel = "gen_ai.response.model";
+        public const string FinishReason = "gen_ai.response.finish_reason";
+        public const string PromptToken = "gen_ai.response.prompt_tokens";
+        public const string CompletionToken = "gen_ai.response.completion_tokens";
+        public const string Prompt = "gen_ai.content.prompt";
+        public const string Completion = "gen_ai.content.completion";
+        public const string Address = "server.address";
+        public const string Port = "server.port";
```

Comment:
```suggestion
        public const string Port = "server.port";
        public const string ErrorType = "error.type";
```

it's the most useful attribute in my small world :)

Reply:
I created a new extension method `SetError` which will attach this attribute to the span.